classes/db/SchemaDAO.inc.php

```diff
@@ -0,0 +1,250 @@
+			}
+		}
+
+		if (empty($primaryDbProps)) {
+			fatalError('Tried to insert ' . get_class($object) . ' without any properties for the ' . $this->tableName . ' table.');
+		}
+
+		$columnsList = join(', ', array_keys($primaryDbProps));
+		$bindList = join(', ', array_fill(0, count($primaryDbProps), '?'));
+		$this->update("INSERT INTO $this->tableName ($columnsList) VALUES ($bindList)", array_values($primaryDbProps));
+
+		$object->setId($this->getInsertId());
+
+		// Add additional properties to settings table if they exist
+		if (count($sanitizedProps) !== count($primaryDbProps)) {
+			$columns = array($this->primaryKeyColumn, 'locale', 'setting_name', 'setting_value');
+			$columnsList = join(', ', $columns);
+			$bindList = join(', ', array_fill(0, count($columns), '?'));
+			foreach ($schema->properties as $propName => $propSchema) {
+				if (!isset($sanitizedProps[$propName]) || array_key_exists($propName, $this->primaryTableColumns)) {
+					continue;
+				}
+				if (!empty($propSchema->multilingual)) {
+					foreach ($sanitizedProps[$propName] as $localeKey => $localeValue) {
+						$this->update("INSERT INTO $this->settingsTableName ($columnsList) VALUES ($bindList)", array(
+							$object->getId(),
+							$localeKey,
+							$propName,
+							$this->convertToDB($localeValue, $schema->properties->{$propName}->type),
+						));
+					}
+				} else {
+					$this->update("INSERT INTO $this->settingsTableName ($columnsList) VALUES ($bindList)", array(
+						$object->getId(),
+						'',
+						$propName,
+						$this->convertToDB($sanitizedProps[$propName], $schema->properties->{$propName}->type),
+					));
+				}
+			}
+		}
+
+		return $object->getId();
+	}
+
+	/**
+	 * Update an object
+	 *
+	 * When updating an object, we remove table rows for any settings which are
+	 * no longer present.
+	 *
+	 * Multilingual fields are an exception to this. When a locale key is missing
+	 * from the object, it is not removed from the database. This is to ensure
+	 * that data from a locale which is later disabled is not automatically
+	 * removed from the database, and will appear if the locale is re-enabled.
+	 *
+	 * To delete a value for a locale key, a null value must be passed.
+	 *
+	 * @param $object DataObject The object to insert into the database
+	 */
+	public function updateObject($object) {
+		$schemaService = ServicesContainer::instance()->get('schema');
+		$schema = $schemaService->get($this->schemaName);
+		$sanitizedProps = $schemaService->sanitize($this->schemaName, $object->_data);
+
+		$primaryDbProps = [];
+		foreach ($this->primaryTableColumns as $propName => $columnName) {
+			if ($propName !== 'id' && isset($sanitizedProps[$propName])) {
+				$primaryDbProps[$columnName] = $this->convertToDB($sanitizedProps[$propName], $schema->properties->{$propName}->type);
+			}
+		}
+
+		$set = join('=?,', array_keys($primaryDbProps)) . '=?';
+		$this->update(
+			"UPDATE $this->tableName SET $set WHERE $this->primaryKeyColumn = ?",
```

Comment:
`$this->primaryKeyColumn` will always be defined in PHP (right?), but are we sure `$set` isn't a possible venue for SQL injections? I'd rather err on the side of escaping, even if it's not a practical risk, just so we don't have to worry about it in the future.

Reply:
(Obviously that comment applies elsewhere too)

Reply:
Good catch. I was thinking the param binding was enough but I think you're right. How would you recommend it be escaped here, knowing that the array `$primaryDbProps` may contain serialized arrays or objects?

Reply:
Looking into this further, I think that we're pretty safe. Here's the full code block:

```
		$primaryDbProps = [];
		foreach ($this->primaryTableColumns as $propName => $columnName) {
			if ($propName !== 'id' && isset($sanitizedProps[$propName])) {
				$primaryDbProps[$columnName] = $this->convertToDB($sanitizedProps[$propName], $schema->properties->{$propName}->type);
			}
		}

		$set = join('=?,', array_keys($primaryDbProps)) . '=?';
		$this->update(
			"UPDATE $this->tableName SET $set WHERE $this->primaryKeyColumn = ?",
			array_merge(array_values($primaryDbProps), array($object->getId()))
		);
```

`array_keys($primaryDbProps)` comes from:

```
foreach ($this->primaryTableColumns as $propName => $columnName) {
  ...
  $primaryDbProps[$columnName] = ...
}
```

And `$this->primaryTableColumns` is defined in each DAO: https://github.com/NateWr/ojs/blob/i3594_forms/classes/journal/JournalDAO.inc.php#L37-L44.

So I don't think there's any way for user-defined data to go into that `$set` variable. We are only pulling in the database column names from the DAO.

Would you still like me to escape those column names? If so, is a simple alphanumeric + `_` regex safe to use for column names?

Reply:
Would you prefer this formulation, which seems to do the same thing in code that is more straightforward about the `SET` string being constructed?

```
$set = $params = [];
foreach ($this->primaryTableColumns as $propName => $column) {
	$set[] = $column . ' = ?';
	$params[] = $this->convertToDb($sanitizedProps[$propName], $schema->properties->{$propName}->type);
}
$this->update("UPDATE site SET " . join(',', $set), $params);
```

Reply:
If you're confident that there's nothing injectable (i.e. it all comes from sources fully in the code), then I think we're OK. There's no good "standard" function for escaping a column or table name AFAIK so better not to invent one. Thanks!